synapse/replication/tcp/handler.py

```diff
@@ -599,6 +612,16 @@ async def _process_position(
         """
         stream = self._streams[stream_name]
 
+        if stream.can_discard_position(
+            cmd.instance_name, cmd.prev_token, cmd.new_token
+        ) and self.is_stream_connected(conn, cmd.stream_name):
+            logger.debug(
+                "Discarding redundant POSITION %s/%s",
+                cmd.instance_name,
+                cmd.stream_name,
+            )
+            return
```

Comment:
As we process the queue the current token for the stream changes, so by the time we get to the point of processing the position we might now be able to discard. I think if anything this is the correct place to do the check, and before is just a fast path to avoid unnecessarily creating a background job.